bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,236 @@
+        &self,
+        scripts: BTreeMap<K, impl Iterator<Item = (u32, Script)>>,
+        stop_gap: Option<usize>,
+        local_chain: &BTreeMap<u32, BlockHash>,
+    ) -> Result<(SparseChain, BTreeMap<K, u32>), ElectrumError> {
+        let mut sparse_chain = SparseChain::default();
+
+        // Check for reorgs.
+        // In case of reorg, new checkpoints until the last common checkpoint is added to the structure
+        for (&existing_height, &existing_hash) in local_chain.iter().rev() {
+            let current_hash = self
+                .client
+                .block_header(existing_height as usize)?
+                .block_hash();
+            sparse_chain
+                .insert_checkpoint(BlockId {
+                    height: existing_height,
+                    hash: current_hash,
+                })
+                .expect("This never errors because we are working with a fresh chain");
+
+            if current_hash == existing_hash {
+                break;
+            }
+        }
+
+        // Insert the new tip
+        let (tip_height, tip_hash) = self.get_tip()?;
+        sparse_chain
+            .insert_checkpoint(BlockId {
+                height: tip_height,
+                hash: tip_hash,
+            })
+            .expect("This never errors because we are working with a fresh chain");
+
+        let mut keychain_index_update = BTreeMap::new();
+
+        let mut reorgred_tx = Vec::new();
+
+        // Fetch Keychain's last_active_index and all related txids.
+        // Add them into the KeyChainScan
+        for (keychain, mut scripts) in scripts.into_iter() {
+            let mut last_active_index = 0;
+            let mut unused_script_count = 0usize;
+            let mut script_history_txid = Vec::<(Txid, TxHeight)>::new();
+
+            loop {
+                if let Some((index, script)) = scripts.next() {
+                    let history = self
+                        .script_get_history(&script)?
+                        .iter()
+                        .map(|history_result| {
+                            if history_result.height > 0
+                                && (history_result.height as u32) < tip_height
+                            {
+                                return (
+                                    history_result.tx_hash,
+                                    TxHeight::Confirmed(history_result.height as u32),
+                                );
+                            } else {
+                                return (history_result.tx_hash, TxHeight::Unconfirmed);
+                            };
+                        })
+                        .collect::<Vec<(Txid, TxHeight)>>();
+
+                    if history.is_empty() {
+                        unused_script_count += 1;
+                    } else {
+                        last_active_index = index;
+                        script_history_txid.extend(history.iter());
```

Comment:
why not just insert into the sparse chain right here.

Reply:
Good point.. Done..